CodeSignal/dockerfiles
# Acceptance-tests Dockerfile

Closes #63.

Creates a new docker image with `puppeteer`, `sinon`, `mocha` and `chai` globally installed.

__How to test this:__

I deployed a new front-end coderunner that you can find on Rundash with IP `192.241.245.30`. SSH into it! Here's what you'll find:
1) If you `cd` into `coderunners` directory, you'll find that it has `Dockerfile` and `test.js`, in addition to the regular files. `Dockerfile` is the file that I added in this PR; `test.js` has the following content:
```js
const puppeteer = require('puppeteer');
const { expect } = require('chai');
const sinon = require('sinon');
const globalVariables = [global.browser, global.expect];

// puppeteer options
const opts = {
  headless: true,
  args: [
    '--no-sandbox',
    '--disable-setuid-sandbox'
  ]
};

describe('sample test', function () {
  let page;

  before(async function () {
    global.expect = expect;
    global.browser = await puppeteer.launch(opts);

    page = await browser.newPage();
    await page.goto('https://www.google.com/');
  });

  it('should have headless chrome of specified version', async function () {
    const browserVersion = await browser.version();
    expect(browserVersion).to.equal('HeadlessChrome/77.0.3844.0');
  });

  it('should have the correct page title', async function () {
    const title = await page.title();
    expect(title).to.eql('Google');
  });

  it('should fail on incorrect page title title', async function () {
    const title = await page.title();
    expect(title).to.eql('Foogle');
  });

  it('should work with sinon', () => {
    const object = { method: function () {} };
    const spy = sinon.spy(object, "method");

    object.method(42);
    object.method(1);

    expect(spy.withArgs(42).calledOnce).to.be.true;
    expect(spy.withArgs(1).calledOnce).to.be.true;
  });

  after(async function () {
    await page.close();

    browser.close();
    global.browser = globalVariables.browser;
    global.expect = globalVariables.expect;
  });
});
```

2) If you run `docker images`, you should see the following:
```
REPOSITORY               TAG                 IMAGE ID            CREATED             SIZE
puppet                   latest              f4bb5424d0e1        9 minutes ago       1.6GB
codesignal/ubuntu-base   v5.0                dc505459ebfd        3 hours ago         1.16GB
codesignal/ubuntu-base   v4.0                dc6b8cfbadc2        6 months ago        1.11GB
codesignal/galen         v1.1                9b799b39bbed        8 months ago        2.03GB
```

`puppet:latest` is the image I built from this new `Dockerfile`.

3) Run `docker run -itv "/root/coderunner":/usercode  puppet:latest /bin/bash` to span a new container using `puppet` image, with `/coderunner` content under the `/usercode` directory in an interactive mode. From there, run something like `mocha usercode/test.js` to get output similar to this one:
```
  sample test
    1) should have headless chrome of specified version
    ✓ should have the correct page title
    2) should fail on incorrect page title title
    ✓ should work with sinon


  2 passing (899ms)
  2 failing

  1) sample test
       should have headless chrome of specified version:

      AssertionError: expected 'HeadlessChrome/77.0.3854.3' to equal 'HeadlessChrome/77.0.3844.0'
      + expected - actual

      -HeadlessChrome/77.0.3854.3
      +HeadlessChrome/77.0.3844.0
      
      at Context.<anonymous> (/usercode/test.js:29:31)
      at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:189:7)

  2) sample test
       should fail on incorrect page title title:

      AssertionError: expected 'Google' to deeply equal 'Foogle'
      + expected - actual

      -Google
      +Foogle
      
      at Context.<anonymous> (/usercode/test.js:39:22)
      at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:189:7)
```

Don't be alarmed by `browser.version` returning `HeadlessChrome...` and not `HeadlessChromium...` or something ! It's [perfectly normal](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#browserversion).

That's it! We can now run tests using puppeteer pointing on both remote and local URLs.
